numpy/core/tests/test_mem_policy.py

```diff
@@ -0,0 +1,286 @@
+                }
+                return (void *)((char *)funcs->realloc(real, sz + 64) + 64);
+            }
+            else {
+                char *real = (char *)funcs->realloc(p, sz + 64);
+                if (real == NULL) {
+                    return NULL;
+                }
+                snprintf(real, 64, "originally allocated "
+                         "%ld  via realloc", (unsigned long)sz);
+                return (void *)(real + 64);
+            }
+        }
+        /* As an example, we use the standard {m|c|re}alloc/free funcs. */
+        static SecretDataAllocatorFuncs secret_data_handler_ctx = {
+            malloc,
+            calloc,
+            realloc,
+            free
+        };
+        static PyDataMem_Handler secret_data_handler = {
+            "secret_data_allocator",
+            {
+                &secret_data_handler_ctx, /* ctx */
+                shift_alloc,              /* malloc */
+                shift_zero,               /* calloc */
+                shift_realloc,            /* realloc */
+                shift_free                /* free */
+            }
+        };
+        '''
+    more_init = "import_array();"
+    try:
+        import mem_policy
+        return mem_policy
+    except ImportError:
+        pass
+    # if it does not exist, build and load it
+    return extbuild.build_and_import_extension('mem_policy',
+                                               functions,
+                                               prologue=prologue,
+                                               include_dirs=[np.get_include()],
+                                               build_dir=tmp_path,
+                                               more_init=more_init)
+
+
+def test_set_policy(get_module):
+    orig_policy_name = np.core.multiarray.get_handler_name()
+
+    a = np.arange(10).reshape((2, 5))  # a doesn't own its own data
+    assert np.core.multiarray.get_handler_name(a) == orig_policy_name
+
+    orig_policy = get_module.set_secret_data_policy()
+
+    b = np.arange(10).reshape((2, 5))  # b doesn't own its own data
+    assert np.core.multiarray.get_handler_name(b) == 'secret_data_allocator'
+
+    if orig_policy_name == 'default_allocator':
+        get_module.set_old_policy(None)  # tests PyDataMem_SetHandler(NULL)
+        assert np.core.multiarray.get_handler_name() == 'default_allocator'
+    else:
+        get_module.set_old_policy(orig_policy)
+        assert np.core.multiarray.get_handler_name() == orig_policy_name
+
+
+def test_policy_propagation(get_module):
+
+    class MyArr(np.ndarray):
+        pass
+
+    orig_policy_name = np.core.multiarray.get_handler_name()
+    a = np.arange(10).view(MyArr).reshape((2, 5))  # a doesn't own its own data
+    assert np.core.multiarray.get_handler_name(a) == orig_policy_name
+
+    b = np.arange(10).view(MyArr).reshape((2, 5))  # b doesn't own its own data
+    assert np.core.multiarray.get_handler_name(b) == 'secret_data_allocator'
```

Comment:
I think this will fail and should be deleted, or replaced with:

```

orig_policy = get_module.set_secret_data_policy()
b = np.arange(10).view(MyArr).reshape((2, 5))  # b doesn't own its own data
assert np.core.multiarray.get_handler_name(b) == 'secret_data_allocator'
get_module.set_old_policy(orig_policy)
```

Reply:
right, thanks